### PR TITLE
Fixes some gas estimates

### PR DESCRIPTION
**Pull request by jacqueswww**
### - What I did

- Added --show-gas-estimates to work with -f ir; this was really helpful in for to figure out what is going on with the gas estimates on the LLL level. So I figured it can be added for more improvement ;)
- Replaced get_contract with get_contract_with_gas_estimation in most places, internal/external call tests abd RLPList are still outstanding, but are perfect for another PR (this one is quite a number of changes already).

### - How I did it

Started replacing get_contract with get_contract_with_gas_estimation. Starting at the LLL, checking the gas estimate tree and then a bit of trial and error on each test cases the subsequently failed.

### - How to verify it
Run the tests, and check out --show-gas-estimates..
image

### - Description for the changelog

Improved gas estimates.

### CODE REVIEW

File: viper/opcodes.py [hunk 1]
```diff
@@ -71,11 +71,11 @@
 }
 
 pseudo_opcodes = {
-    'CLAMP': [None, 3, 1, 45],
+    'CLAMP': [None, 3, 1, 45 + 25],
```
Comment (DavidKnott) [inline viper/opcodes.py]: Why not just change it to 70?

File: viper/opcodes.py [hunk 2]
```diff
@@ -71,11 +71,11 @@
...
     'UCLAMPLT': [None, 2, 1, 25],
     'UCLAMPLE': [None, 2, 1, 30],
     'CLAMP_NONZERO': [None, 1, 1, 19],
-    'ASSERT': [None, 1, 0, 20],
+    'ASSERT': [None, 1, 0, 20 + 65],
```
Comment (DavidKnott) [inline viper/opcodes.py]: And 85?

File: viper/parser/parser_utils.py [hunk 3]
```diff
@@ -165,6 +170,10 @@ def repr(self):
         o = ''
         if self.annotation:
             o += '/* %s */ \n' % self.annotation
+        if self.repr_show_gas and self.gas:
+            OKBLUE = '\033[94m'
+            ENDC = '\033[0m'
+            o += OKBLUE + "{" + ENDC + str(self.gas) + OKBLUE + "} " + ENDC  # add gas for info.
```
Comment (DavidKnott) [inline viper/parser/parser_utils.py]: Could you explain what this does?

File: viper/parser/parser_utils.py [hunk 4]
```diff
@@ -238,9 +247,13 @@ def make_byte_array_copier(destination, source):
         raise TypeMismatchException("Cannot cast from greater max-length %d to shorter max-length %d" % (source.typ.maxlen, destination.typ.maxlen))
     # Special case: memory to memory
     if source.location == "memory" and destination.location == "memory":
-        return LLLnode.from_list(
+        gas_calculation = opcodes.GIDENTITYBASE + \
+            opcodes.GIDENTITYWORD * (ceil32(source.typ.maxlen) // 32)
+        o = LLLnode.from_list(
```
Comment (DavidKnott) [inline viper/parser/parser_utils.py]: I love how you're calculating the memory usage.

File: viper/opcodes.py (hunk 1 above)
Comment (jacqueswww) [reply viper/opcodes.py]: Sure will do so, I just kept it separate at the time to know the delta easily. Forgot to change it to a single value hehe ;)

File: viper/parser/parser_utils.py (hunk 3 above)
Comment (jacqueswww) [reply viper/parser/parser_utils.py]: This just for the nice blue coloured braces one sees on the `--show-gas-estimates` output. I will also add a test for it ;)

File: viper/parser/parser_utils.py (hunk 4 above)
Comment (jacqueswww) [reply viper/parser/parser_utils.py]: :smile: thanks! I picked it from the pyethereum specials.py file. We can now do it in other known precompiled contract in the same way using `add_gas`.

**Comment by DavidKnott** [comment]
LGTM

**Pull request merged**